### PR TITLE
Addition of `factor_cache`

### PR DESCRIPTION
**Pull request by haru-44**
A new class, `FactorCache`, was created to cache prime factors. When `factorint` is executed, it caches the found prime factors for use in subsequent factorizations.

#### References to other Issues or PRs

#27124 

#### Release Notes

* ntheory
  * The `factor_cache` has been added to store the results of prime factorization. Results from `factorint` are automatically cached, and users can also add their own entries to the cache.

### CODE REVIEW

**Comment by smichr** [comment]
I like the way that facators are added for a known number. But should there be a check that all factors have been added? e.g. would `add(2*3*5*7, [3, 7])` be an error?

**Comment by haru-44** [comment]
I haven’t checked if all prime factors are included. Even if only some of the prime factors are known, I think it should be acceptable. The function `add(2*3*5*7, [3, 7])` is expected to add the following two items to the cache:
- `2*3*5*7`: 7
- `2*3*5`: 3

On the other hand, if the factors provided are not prime, it will throw an error. For example, `add(2*3*5*7, [6])` will result in an error. Although non-prime factors could still be useful information, I decided against handling them to keep the processing simpler.

File: sympy/ntheory/factor_.py [hunk 1]
```diff
@@ -608,6 +611,117 @@ def _factors():
...
+            self.add(n, factors)
+            return self._cache[n]
+        return default
+
+    def add(self, n: int, factors: list[int]) -> None:
+        for p in sorted(factors, reverse=True):
+            self[n] = p
+            n, _ = remove(n, p)
+
+    def get_external(self, n: int) -> list[int] | None:
```
Comment (sylee957) [inline sympy/ntheory/factor_.py:718]: I notice that this function is not implemented before merge

Comment (haru-44) [reply sympy/ntheory/factor_.py:718]: The `get_external` function is meant to be implemented by the user and is not intended to be provided by SymPy. If the user does not implement it, the function will always return `None`, and lines 709 and 710 will never be executed.

Does this explanation work for you? I was not entirely sure about the intention behind "before merge."

Comment (sylee957) [reply sympy/ntheory/factor_.py:718]: No, I don't think that it is possible for user to override get_external to be working.
I think that it is only possible with overriding, but sympy uses an instance of FactorCache so there is no option to override get_external for that function.

Maybe it is possible with factor_cache.get_external = ... but the type annotation could be different

Comment (sylee957) [reply sympy/ntheory/factor_.py:718]: Okay, I notice that it is documented

Comment (sylee957) [reply sympy/ntheory/factor_.py:718]: But anyway, having a in-place replacement of method with function with `self` parameter looked quite confusing to me

File: sympy/ntheory/factor_.py [hunk 2]
```diff
@@ -608,6 +611,117 @@ def _factors():
...
+    def maxsize(self, value: int | None) -> None:
+        if value is not None and value <= 0:
+            raise ValueError("maxsize must be None or a non-negative integer.")
+        self._maxsize = value
+        if value is not None:
+            while len(self._cache) > value:
+                self._cache.popitem(False)
+
+    def get(self, n: int, default=None):
+        """ Return the prime factor of ``n``.
```
Comment (Quuxplusone) [inline sympy/ntheory/factor_.py:699]: Coming here from the user-facing documentation: It's unclear to me what "_the_ prime factor of `n`" means. Source-diving a little bit, it looks like the cache holds only one prime factor per input number; to get the full factorization of the input number you'd have to call `sympy.factor_cache.get` in a loop... except that that's what `sympy.ntheory.primefactors(n)` already does for you, so the user should never actually have to interact with `sympy.factor_cache` except perhaps to do

    sympy.factor_cache = sympy.FactorCache(maxsize=100000)

at the start of their program. Is that basically correct?

And if so, this docstring should be updated to give the user some of this information. Simply replacing "Return the prime factor of `n`" with "Return one prime factor of `n`" or "Return the largest prime factor of `n` previously memoized by the cache" would be a big improvement.

Comment (haru-44) [reply sympy/ntheory/factor_.py:699]: As you pointed out, "Return the prime factor of `n`" is somewhat imprecise. "Return the largest prime factor of `n` previously memoized by the cache" would be more accurate, though I'm not sure if it is the best wording.

**Pull request merged**